auth/server/base/src/main/java/org/wildfly/security/cache/CachedIdentity.java

```diff
@@ -64,11 +70,45 @@ public CachedIdentity(String mechanismName, boolean programmatic, Principal prin
         this(mechanismName, programmatic, null, principal);
     }
 
+    /**
+     * Creates a new instance based on the given <code>mechanismName</code> and <code>principal</code>.
+     *
+     * @param mechanismName the name of the authentication mechanism used to authenticate/authorize the identity
+     * @param programmatic indicates if this identity was created as a result of programmatic authentication
+     * @param principal the principal of this cached identity
+     * @param roles the roles assigned to this cached identity
+     */
+    public CachedIdentity(String mechanismName, boolean programmatic, Principal principal, Set<String> roles) {
+        this(mechanismName, programmatic, null, principal, roles);
+    }
+
     private CachedIdentity(String mechanismName, boolean programmatic, SecurityIdentity securityIdentity, Principal principal) {
         this.mechanismName = checkNotNullParam("mechanismName", mechanismName);
         this.programmatic = programmatic;
         this.name = checkNotNullParam("name", checkNotNullParam("principal", principal).getName());
         this.securityIdentity = securityIdentity;
+        if (securityIdentity != null && securityIdentity.getPrincipal() != null) {
+            this.roles = rolesToSet(securityIdentity.getRoles());
+        } else {
+            this.roles = Collections.emptySet();
+        }
+    }
+
+    private CachedIdentity(String mechanismName, boolean programmatic, SecurityIdentity securityIdentity, Principal principal, Set<String> roles) {
+        this.mechanismName = checkNotNullParam("mechanismName", mechanismName);
+        this.programmatic = programmatic;
+        this.name = checkNotNullParam("name", checkNotNullParam("principal", principal).getName());
+        this.securityIdentity = securityIdentity;
+        this.roles = roles;
+    }
+
+    private Set<String> rolesToSet(Roles roles) {
+        Iterator<String> iterator = roles.iterator();
+        Set<String> resultSet = new HashSet<>();
+        while (iterator.hasNext()) {
+            resultSet.add(iterator.next());
+        }
+        return resultSet;
```

Comment:
Given that this set will be generated per user and will typically be small in size, I might suggest using a more lightweight data structure (e.g. TreeSet) and leverage empty and singleton implementations if possible.
e.g.
```java
    private Set<String> rolesToSet(Roles roles) {
        if (roles.isEmpty()) return Set.of();
        Iterator<String> iterator = roles.iterator();
        String role = iterator.next();
        if (!iterator.hasNext()) return Set.of(role);
        Set<String> result = new TreeSet<>();
        result.add(role);
        while (iterator.hasNext()) {
            result.add(iterator.next());
        }
        return result;
    }
```
Perhaps this logic could reside within a Roles.toSet() method?

Reply:
@pferraro Good points! I have moved the method to the Roles class and I have used your `TreeSet` approach. I just modified it slightly to not use `Set.of` because we have to use Java 8 APIs. Thank you!